src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs

```diff
@@ -423,274 +432,241 @@ public void Abort()
             // TODO
         }
 
-        public void StartReadingRequestBody()
+        public void StartProcessingRequestAndResponseBody()
         {
-            if (_readingTask == null)
+            if (_readWriteTask == null)
             {
-                _readingTask = ProcessRequestBody();
+                _readWriteTask = ProcessRequestAndResponseBody();
             }
         }
 
-        private async Task ProcessRequestBody()
+        // IIS can only handle on async operation per request at a time.
+        // To guarantee this, we have a single loop to read and write the request/response
+        private async Task ProcessRequestAndResponseBody()
         {
+            var doneReading = false;
             try
             {
                 while (true)
                 {
-                    // These buffers are pinned
-                    var wb = Input.Writer.GetMemory(MinAllocBufferSize);
-                    _inputHandle = wb.Retain(true);
-
-                    try
+                    if (!doneReading && _allowedToRead)
```

Comment:
@halter73 I think having a single loop is fundamentally flawed. If I want to read the entire request body before writing, how can the reads unblock the writes with a single loop? Calling await WriteFromOutputPipe() will block until data arrives, which may not happen until the response body is read.